### ISSUE TITLE
Type narrowing is not allowed if variable is reassigned after the use site inside if block

### ISSUE THREAD

**Issue by vladima** [Microsoft/TypeScript#1764]
Current implementation is very conservative when looking for assignments during application of type narrowing.
```typescript
function len(x: number | string) {
	 if (typeof x === 'number') {
		 var t = x.toExponential();
		 //x = "aaaa"; // uncomment me
	 }
}
```

variable is reassigned after the usage of narrowed type but type narrowing still not kicked in

**Comment by ahejlsberg** [comment]
We'd have to do flow analysis to properly reason about this. That might be worthwhile (and would allow us to detect unreachable code, unassigned variables, etc. as well), but we need to consider the impact on complexity and performance.

**Comment by vladima** [comment]
btw, there is already a PR for control flow analysis: #1287

**Comment by jbondc** [comment]
Control flow looks really nice :+1:

**Comment by mhegazy** [comment]
So what is the proposal?

**Comment by ivogabe** [comment]
I've opened a PR #6959 that will fix this.

**Comment by ivogabe** [comment]
I think that this can be closed, #8010
